[jamal.park] 지하철 서비스 Step3 리뷰 요청드립니다.

jypark5826 opened the pull request:
안녕하세요, 뉴크루 자말입니다.
지하철 서비스 Step3 리뷰 요청드립니다.
감사합니다:)

waylon-n committed:
docs: README.md 구현 기능 작성

waylon-n committed:
feat: 모든 Section 탐색

SectionDao에서 모든 Section을 불러와 Sections로 맵핑

waylon-n committed:
feat: 최단 경로 및 거리 탐색

jgrapht 라이브러리 사용
불러온 모든 Section을 이용해 모든 Station의 id를 정점으로 사용
모든 Section 정보를 간선으로 사용해 WeightedMultigraph 생성 및 최단 경로 탐색
PathResponse로 반환

waylon-n committed:
refactor: Vertex, Edge 추가 메소드 분리

waylon-n committed:
docs: 구현 기능 목록 업데이트 (step2)

waylon-n committed:
feat: 토큰 발급 및 로그인

유효 로그인 검사
로그인 시 액세스를 위한 토큰 발급
정보 관련 기능(/members/me) 요청 시 토큰 검사

waylon-n committed:
feat: FAVORITE 테이블 스키마

waylon-n committed:
feat: Favorite 추가 기능

waylon-n committed:
feat: AuthenticationPrincipalArgumentResolver Configuration

waylon-n committed:
feat: 즐겨찾기 목록 조회

waylon-n committed:
feat: 즐겨찾기 삭제

waylon-n committed:
docs: 구현 기능 목록 업데이트 (step3)

waylon-n committed:
feat: 노선 추가 요금 입력, 기본값 0

waylon-n committed:
refactor: PathExplorer 최단 경로 탐색 유틸 생성

waylon-n committed:
feat: 요금 계산

거리, 노선, 나이에 따른 요금 계산 추가
거리, 나이 구간에 enum 사용

waylon-n committed:
feat: 연관 데이터 삭제

노선 삭제 시 관련된 구간 모두 삭제
특정 구간이 포함된 역 삭제 불가
특정 역 삭제 시 관련된 즐겨찾기 모두 삭제

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,97 @@
+        return PathResponse.of(shortestPath);
+    }
+
+    public Path getShortestPath(LoginMember loginMember, Station source, Station target) {
```
최단거리경로를 잘 구해주셨습니다. 👍 
여기에 역할 분리를 좀 더 진행해주면 더 좋은 코드가 될 거 같습니다. ☺️

1. getShortestPath() 매소드가 하는 일이 많습니다.
최단경로구하기,  요금계산(추가요금, 연령별할인요금) 로직을 매소드에 모두 구현하고 있습니다.  로직을 역할에 맞게 분리하여 확장성과 가독성을 높여 주세요.

 2. 요청시 마다 WeightedMultigraph 를 생성하여 최단경로를 구하는 것은 비용이 많이 드는 작업입니다.  서비스에서 역 추가나 노선 변경작업은 자주 일어나지 않습니다.  그렇다면 최단경로를 구해주는 객체는 한번 초기화 후 계속 사용하는 것이 효과적 입니다. 최단경로구하기 역할을 분리하면서 같이 개선해 보시기 바랍니다. 😊

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,97 @@
+        if (loginMember != null) {
+            AgeFare ageFare = AgeFare.getAgeFare(loginMember.getAge());
+            fare -= ageFare.calculateDiscountFareByAge(fare);
+        }
```
요금계산에 필요한 최단이동거리, 노선별 추가요금, 이용자의 연령 항목을 input 으로 주면 요금을 알려주는 객체를 만들어 주면 로직을 분리해 낼 수 있습니다. 자세한 요금정책(거리별요금정책, 추가요금정책, 연령별 요금정책) 에 대한 책임은 해당 객체가 책임을 가져 가도록 하면 로직이 제거 될 거 같네요 ☺️

sehan commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -1,13 +1,18 @@
 
+import javax.servlet.http.HttpServletRequest;
+
+@Component
```
AuthenticationPrincipalConfig 에서 AuthenticationPrincipalArgumentResolver 를 Bean 으로 등록해 주었기 때문에 @Component 를 해주지 않아도 괜찮습니다.

sehan commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,40 @@
 
 @RestController
+@RequestMapping("favorites")
 public class FavoriteController {
```
즐겨찾기 기능은 인증된 사용자의 요청만 허용해야 합니다.
현재 인증을 위한 구현이 누락되어 있습니다.  
아래 강의 내용에 언급되었던 Interceptor 를 사용한 접근권한 구현이 필요합니다.
https://edu.nextstep.camp/s/bb4PXLji/ls/MIOwm2me

sehan commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,71 @@
+        return new Favorite(lineId, favorite.getMemberId(), favorite.getSourceStation(), favorite.getTargetStation());
+    }
+
+    public List<Favorite> findById(Long memberId) {
```
findById 란 이름은 일반적으로 목록(List) 이 아니라 하나의 객체를 return 할 때 사용합니다. 
findByMemberId 로 변경 하는것이 좋겠습니다. ☺️

sehan reviewed (changes requested):
안녕하세요. 지하철 서비스 리뷰를 맡은 니오입니다. ☺️

경로조회, 요금계산 기능을 빠르게 잘 구현해주셨습니다. 👏👏👏
기능구현 측면에서는 인증부분에 대한 구현이 누락된 것으로 확인됩니다. 로그인(토근발급) 까지는 구현이 되어 있는데  인증이 요구되는 API 에 권한체크 부분이 보이지 않습니다. 
그 외에 경로조회, 요금계산 부분에 비지니스 로직이 하나의 매소드에 집중되어 있어서 역할의 분리가 필요해 보입니다. 관련해서 코멘트를 남겨두었습니다.  코멘트 확인 해주세요 ☺️

수정요청 드립니다.

jamal-park committed:
refactor: 6세 미만 할인 요금 추가

jamal-park committed:
refactor: component 어노테이션 제거

jamal-park committed:
refactor: FavoriteDao 메소드명 변경

jamal-park committed:
refactor: 즐겨찾기 인증 관련

유효하지 않은 토큰으로 즐겨찾기 요청을 보낸 경우 테스트 추가
인터셉터로 토큰 유효성 검증

jamal-park committed:
refactor: 경로조회, 요금계산 로직

PathService 로직 분리
StationGraph, FareCalculator 생성

jypark5826 commented:
피드백 주신 내용 바탕으로 리팩토링 해보았습니다.
첫 번째 코멘트 관련해서는 객체를 한번 초기화 후 계속 사용할 방법이 떠오르지 않아서 일단 로직 분리만 해놓았습니다ㅜ
다시 피드백 부탁드립니다!

sehan commented on src/main/java/subway/AuthenticationPrincipalConfig.java:
```diff
@@ -25,4 +27,16 @@ public void addArgumentResolvers(List argumentResolvers) {
     }
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
```
Interceptor 를 사용해서 접근권한제어 처리 잘 해 주셨습니다. 👍 

한가지 덧 붙이자면, LoginInterceptor 객체는 @Bean 으로 등록할 필요는 없기 때문에 addInterceptors() 내에서 바로 생성해서 추가 해 주면 됩니다.
```
registry.addInterceptor(new LoginInterceptor(authService))
```

sehan commented on src/test/java/subway/favorite/FavoriteAcceptanceTest.java:
```diff
@@ -71,6 +73,33 @@ void manageMember() {
         즐겨찾기_삭제됨(deleteResponse);
     }
 
+    @DisplayName("유효하지 않은 토큰으로 즐겨찾기 기능에 접근한다.")
```
acceptance test 코드 작성까지 아주 좋습니다. 👍

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -38,60 +31,13 @@ public PathResponse findShortestPathResponse(LoginMember loginMember, Long sourc
-        addSectionEdge(graph, lines);
+        List<Station> stations = dijkstraShortestPath.getPath(source, target).getVertexList();
+        int distance = (int) dijkstraShortestPath.getPathWeight(source, target);
+        int fare = FareCalculator.calculate(distance, stations, lines, loginMember);
```
요금계산 로직을 멋지게 분리 하셨네요 💯 💯 💯

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -38,60 +31,13 @@ public PathResponse findShortestPathResponse(LoginMember loginMember, Long sourc
-        addStationVertex(graph, lines);
-        addSectionEdge(graph, lines);
+        List<Station> stations = dijkstraShortestPath.getPath(source, target).getVertexList();
+        int distance = (int) dijkstraShortestPath.getPathWeight(source, target);
```
StationGraph 객체로 jgrapht 를 초기화하는 역할만을 분리했습니다.  분리한 것에 일단 박수를 보냅니다. 👏👏👏

이제 첫발을 내딛었으니 StationGraph 에 부여하면 더 좋을 것들을 생각해 봅시다.  StationGraph 객체는 현재 jgrapht 에 의존적인 매소드들을 가지고 있습니다. 대표적으로 getDijkstraShortestPath() 입니다. 외부 라이브러리에 의존적인 매소드명은 해당 라이브러리에 의존적인 코드를 재생산하게 되어 좋지 않습니다. 
먼저 StationGraph 객체의 역할을 명확해 해줄 필요가 있습니다. 결국 StationGraph 에서 DijkstraShortestPath 를 꺼내서 최단이동경로와 이동거리를 구하는 것이 목적입니다. 그렇다면 StationGraph 가 public 으로 제공해야 하는 메소드는 getShortestPath() 면 적합해 보입니다.  jgrapht 에 의존적이지도 않고 역할에 부합하는 이름입니다. 
 그럼 getShortestPath() 가 return 해야 하는 타입은 무엇일까요? 최단이동경로와 이동거리 상태를 포함하는 DTO 객체를 돌려주면 됩니다. 자 그럼 리팩토링 진행 볼 수 있겠죠? ☺️

StationGraph 객체의 이름도 새로운 매소드를 부여함에 따라 그에 맞도록 rename 해주는 것도 좋겠네요 😊

```
interface StationGraph {
    Path getShortestPath();
}

interface Path {
  List<Station> getStations();
  int getDistance();
}
```

sehan commented on src/main/java/subway/path/domain/DistanceFare.java:
```diff
@@ -0,0 +1,33 @@
+
+    OVER_FIFTY_KM(51, 8, 100, 2050),
+    OVER_TEN_KM(11, 5, 100, 1250),
+    DEFAULT(1, 10, 1250, 0);
```
기본운임에 거리별추가요금을 선반영 시켜놨군요 😄 
저는 왜 이생각을 이제서야 하게 된걸까요? 👍👍👍

sehan commented on src/main/java/subway/path/domain/FareCalculator.java:
```diff
@@ -0,0 +1,47 @@
+
+public class FareCalculator {
+    public static int calculate(int shortestDistance, List<Station> stations, List<Line> lines, LoginMember loginMember) {
+        DistanceFare distanceFare = DistanceFare.getDistanceFare(shortestDistance);
+        int fare = distanceFare.calculateExtraFareByDistance(shortestDistance);
+
+        fare += getMaxExtraFareOfLine(stations, lines);
+
+        if (loginMember != null) {
+            AgeFare ageFare = AgeFare.getAgeFare(loginMember.getAge());
+            fare -= ageFare.calculateDiscountFareByAge(fare);
+        }
+
+        return fare;
```
FareCalculator 로 요금계산로직을 분리한 것 아주 좋습니다.

좀 더 개선하면 더 좋을거 같아 요구사항을 더 이야기 해보겠습니다. 👌
개선 되었으면 하는 포인트 입니다.
1. 계산을 위해 항상 전체노선(lines)을 입력받아야 합니다. 노선별 추가요금정책에 따라 요금을 책정할때 전체노선(lines)을 상태로 관리하고 있다면 계산을 위해 항상 입력받지 않아도 됩니다. 물론 전체노선이 필요하기 보다는 실제로는 구간(A->B) 에 해당하는 노선추가요금정보 상태를 관리하는 것이 중요합니다. 
2. 연령별 요금계산에 필요한 정보는 loginMember 가 아니라 age 입니다. calculate() 인자로 필요한 값을 인자로 받는것이 좋습니다.
3. 모든 메소드가 static 으로 구현되어 있습니다. 하지만 앞서 말한 1번 항목을 고려 해보면 static 을 제거 하고 상태를 가지는 객체로 만들어도 괜찮아 보입니다. 

**개선팁**
요금정책들이 3가지 있습니다. 거리별요금, 노선구간별추가요금, 연령별 할인요금.
FareCalculator 가 요금정책들의 상태를 관리하고 calculate() 시 관리하고 있는 요금정책들을 적용하여 요금을 계산해 주면 어떨까요?
앞서 이야기한 내용들을 interface 로 설계 한다면 아래와 같은 예시를 드릴 수 있을거 같습니다. 
```
interface FareCalculator {

    List<FarePolicy> getFarePolicies()
    // Path 가 List<station> stations , int distance 상태를 가진다면
    int calculate( Path path, int age )

}

interface FarePolicy {
    int apply(int fare, Path path, int age)
}
```

sehan commented on src/main/java/subway/path/domain/StationGraph.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+public class StationGraph {
```
구현에 어려움을 이야기 하셨던 부분에 대해서 답을 드립니다. 

"StationGraph 를 @Bean 으로 등록해서 사용하면 경로탐색시마다 graph 를 새로 초기화 하기 않고 재사용 할 수 있을 것이다." 란 맥락이 었는데요.  Spring 에서는 @Configuration 을 통해 @Bean 으로 등록할 객체들을 손쉽게 Bean 으로 등록해서 필요한 곳에 DI 하여 사용할 수 있습니다.  이 방법을 사용해서 StationGraph 를 @Bean 으로 등록하고 PathService 에서는 DI 받아서 사용해 볼 것을 의도한 것입니다. 

현재 코드기반으로 작성한다면 다음과 같습니다.
```
@Configuration
class SubwayApplicationConfiguration {

    @Bean
    public StationGraph stationGraph(LineDao lineDao){
       return new StationGraph(lineDao.findAll())
    }

}

class PathService{
    private StationGraph stationGraph;
    public PathService(StationGraph graph);
}
```

하지만, 위와 같이 작성하게 되면 line 에 section 이 추가/삭제/수정 되거나 line 이 추가/삭제 되면 stationGraph 의 graph 정보도 함께 갱신해줘야 합니다. 그런 점에서 구현난이도가 높아집니다. 
이 부분 리팩토링을 바로 진행하기보다는 다른 코멘트로 남겨놓은 부분을 먼저 진행하고 마지막에 여력이 될때 한번 고민해 보면 좋을거 같네요 😊

sehan reviewed (changes requested):
의도했던 방향으로 피드백을 아주 잘 반영해 주셨습니다. 💯 💯 💯 

추가로 개선되면 좋을 포인트들이 있어서 코멘트 추가로 남겼습니다. 
코멘트 확인 후 수정부탁드려요. 
한번 더 수정요청드립니다.

jypark5826 committed:
refactor: 경로 관련 리팩토링

StationGraph -> PathExplorer
PahExplorer에 getShortestPath 메소드를 추가하여 Path를 리턴
Path 객체에서 fare를 분리
위 변경사항에 맞추어 PathService 수정

jypark5826 committed:
refactor: FareCalculator 리팩토링

FarePolicy 적용
path, fare 패키지 분리

jypark5826 committed:
refactor: 불필요 코드 정리

jypark5826 commented:
꼼꼼한 피드백 정말 감사드립니다:)
피드백 주신 내용 바탕으로 다시 리팩토링 진행해보았습니다.
마지막 코멘트 내용까지 구현하면 너무 늦어질 것 같아서 그 전까지의 내용 먼저 리뷰 요청드립니다!

sehan commented on src/main/java/subway/path/domain/fare/FareCalculator.java:
```diff
@@ -0,0 +1,30 @@
+	private final int INITIAL_FARE = 0;
+	private List<Line> lines;
+
+	public FareCalculator(List<Line> lines) {
+		this.lines = lines;
+	}
+
+	private List<FarePolicy> getFarePolicies() {
+		return Arrays.asList(new DistanceFarePolicy(), new LineFarePolicy(lines), new AgeFarePolicy());
+	}
```
lines 를 멤버로 가지고 있는데 List<FarePolicy> policies 를 멤버로 가지는 것이 더 적합해 보입니다.
생성자에서 바로 policy 들을 생성해서 상태로 유지하는것이 좋겠네요

sehan commented on src/main/java/subway/path/domain/path/PathExplorer.java:
```diff
@@ -0,0 +1,45 @@
+        });
+    }
+
+    public Path getShortestPath(Station source, Station target) {
+    	DijkstraShortestPath dijkstraShortestPath = new DijkstraShortestPath(graph);
+    	List<Station> stations = dijkstraShortestPath.getPath(source, target).getVertexList();
+        int distance = (int) dijkstraShortestPath.getPathWeight(source, target);
+        
+        return new Path(stations, distance);
+    }
```
👍

sehan reviewed (approved):
피드백 반영 아주 잘해주셨습니다. 👍 👍 👍 

요금관련 로직, 경로관련 리팩토링도 잘 진행해 주셨습니다.  
추가된 요금정책도메인에 대한 단위테스트 까지 잘 추가해주셨습니다. 

PR 은 merge 하도록 하겠습니다.  고생하셨습니다. 

덧, 마지막 코멘트 내용은 경로조회 성능개선 측면에서 이후 고민해 보시기 바랍니다. 
남은 한주도 성장하는 소중한 시간 보내길 바랍니다.
